### PR TITLE
Document experiment provider plugin

### LINKED ISSUE #7639: [DOC-FIX] Experiment context provider
### Willingness to contribute

Yes. I can contribute a documentation fix independently.

### URL(s) with the issue

It appears there's a somewhat undocumented plugin interface for experiment context providers. These appear to allow auto-tagging/custom creation of experiments. I see it marked as developer stable.

https://github.com/mlflow/mlflow/blob/branch-1.13/mlflow/tracking/context/abstract_context.py

Are these safe to subclass in a plugin? I see the databricks provider in the source and would like to implement something similar via a plugin. 

### Description of proposal (what needs changing)

Document the experiment context provider and maybe add an example plugin. I'd be willing to contribute it if this is the intended use case for this interface.

### PR DESCRIPTION
**Pull request by jmahlik**
## Related Issues/PRs

Closes #7639 

## What changes are proposed in this pull request?
Document the experiments provider plugin. 

## How is this patch tested?

N/A docs only
- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Document default experiment provider plugin.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: docs/source/plugins.rst [hunk 1]
```diff
@@ -168,6 +173,14 @@ plugin:
        within the ``mlflow_test_plugin`` module) to register.
      - `GitRunContext <https://github.com/mlflow/mlflow/blob/branch-1.13/mlflow/tracking/context/git_context.py#L38>`_,
        `DefaultRunContext <https://github.com/mlflow/mlflow/blob/branch-1.13/mlflow/tracking/context/default_context.py#L41>`_
+   * - Plugins for specifying custom experiment id lookup/creation at runtime. The plugin must have a way to create or look up the experiment id.
+     - mlflow.default_experiment_provider
+     - The entry point name is unused. The entry point value (e.g. ``mlflow_test_plugin.default_experiment_provider:PluginDefaultExperimentProvider``) specifies a custom subclass of
+       `mlflow.tracking.default_experiment.abstract_context.DefaultExperimentProvider <https://github.com/mlflow/mlflow/blob/branch-2.2/mlflow/tracking/default_experiment/abstract_context.py#L6>`_
```
Comment (BenWilson2) [inline docs/source/plugins.rst:179]: I just realized that the existing links point to really old release branches. Could we update the implementation guidance references from <branch-x.x> to `master`?

Comment (dbczumar) [reply docs/source/plugins.rst:179]: If linking to `master`, can we avoid including line numbers as well, since they'll change

Comment (jmahlik) [reply docs/source/plugins.rst:179]: I can link to the files, but over the years they'll 404. Which, actually might not be a bad thing. Thoughts on updating to `"py:class` type of sphinx refs? That way, if someone changes the class, the docs break and they have to update it. 

Just need to install the plugin as part of the docs ci stage.

Comment (jmahlik) [reply docs/source/plugins.rst:179]: ^ if someone can point me to where the docs are generated in ci, will take care of it.

Comment (BenWilson2) [reply docs/source/plugins.rst:179]: @jmahlik the docs build CI config is here: https://github.com/mlflow/mlflow/blob/master/.circleci/config.yml

Comment (BenWilson2) [reply docs/source/plugins.rst:179]: Any new plugins would just need to be added here: https://github.com/mlflow/mlflow/blob/master/requirements/doc-requirements.txt

**Comment by jmahlik** [comment]
Docs are being re-written, going to close this.

**Pull request closed without merging**